### DISCUSSION TITLE
Quarto Website Filters?

### DISCUSSION THREAD

**Discussion by jlgraves-ubc** [quarto-dev/quarto-cli#3285] [Q&A] [answered]
Hi all,

I'm trying to make a filter (i.e. an extension) that adds some custom elements to the `navbar` in Quarto.

However, I'm struggling since I don't seem to be able to get Pandoc/Lua to identify the `div` elements or `ul` elements that contain the navbar as `Div` elements in the AST - they are being identified as `RawBlock` elements.  This means I can't do anything to them.

Is there something special about the website project time that might be doing this?  That I could adjust or change?

Thanks!

**Comment by dragonstyle** [comment, marked as answer]
Hi - sorry for the trouble! We add the navbar in a post processor, so unfortunately it isn't visible in LUA AST. There isn't a great way to get at it aside from javascript right now, though in the long run we would like to give you access to post processors as well (not sure when we'll get there).

In the meantime is there something specific we should consider adding to navbars to support your scenario more directly?

**Reply by jlgraves-ubc** [reply]
Yes!  I have been using JupyterBook for a while, and there is exactly one feature it has which doesn't exist in Quarto: a "launch notebook on hub" button.  The QuantEcon theme has the best example of it:

https://datascience.quantecon.org/

(The play arrow at the top)

But the standard theme is also fine:

https://comet.arts.ubc.ca/docs/econ_intro/intro_jupyter/intro_to_jupyter.html

(The rocketship)

This is incredibly valuable, mainly for teaching and textbooks, since it allows one-click from reading to writing code.  My hope was that with the more flexible frontmatter for Quarto projects you could configure it flexibly as:

```yaml
hub-selector:
  position:
    location: navbar
    alignment: left
  repo: "repo_URL"
  branch: "main"
  file: "path/to/file.ipynb"
  app: lab
  hubs:
  - name: "Display Name"
    type: [jupyterhub, collab, binder]
    url: URL_to_hub
  - name: "Other Display Name"
    type: collab
    url: URL-to-hub
```

My original extension plan was to use a filter to grab the `<ul>` element for the navbar and drop in a new element with some pop-up functionality.  My fallback plan was just to create a menu with `nbgitpuller` links pre-defined based on the YAML like:

```lua
local function nbGitPuller(repo, branch, url, app, file)
  -- Generate an nbgitpuller link for the hub, repo, and branch
  -- All of the inputs should be strings before being passed to the file
  
    local path = "tree%2F"
    if app == "lab" then
      path = "lab%2Ftree%2F"
    end
  
    repo = urlencode(repo)
    path = path .. urlencode(file)  -- need to add handler for no file
    branch = urlencode(branch)
  
    local nburl = url .. "/hub/user-redirect/git-pull?repo=" .. repo .. "&urlpath=" .. path .. "&branch=" .. branch
  
    return nburl
  
  end
```

This didn't seem to work either - I could update the `Meta` appropriately, but it wasn't reflected in the navbar, likely due to the ordering of the post-processing for the object or something.

I know I could it it with a shortcode, but it would look kind of bad, and should live in the navbar.  This is literally the only feature from JupyterBooks I think Quarto doesn't do better.

**Reply by dragonstyle** [reply]
This is really excellent and timely feedback - thank you so much for taking the time to write it up. It is timely in part because I am working on improving our support for Notebooks (as an output artifact that is accessible from other formats) right now! We're envisioning being able to produce computational notebooks as artifacts accessible from web pages (e.g. make a notebook of a page, or even make notebooks for sections) and automatically generating / linking to them. I anticipate that we're still a little ways away from 1.3 builds, but I think these features will be present in early 1.3 builds and I'll do my best to remember to get back to this thread when that is available so we can hear your thoughts / feedback on the early experience.

**Reply by jlgraves-ubc** [reply]
That's great!  I use Quarto a lot right now for teaching, so let me know and I'll be happy to help out in whatever way I can.

**Reply by jav-ed** [reply]
Hi, I am doing some research with regard, what is the better choice jupyter books or quarto. You have stated quarto would be better, would you mind providing some arguments - thank you :)

**Reply by jlgraves-ubc** [reply]
It's not "better" in an absolute sense, but it does do basically everything JupyterBook does, at the cost of a little more overhead.

* Quarto is more flexible than JupyterBook, allowing more types of sites with different features and themes, but requires more configuration which takes time to set up and tweak.
* JupyterBook is more tightly integrated with JupyterNotebook formats, which allows it to do some things that Quarto currently doesn't do (see the above discussion), but on the other hand it means that it's less capable with other formats and relies heavily on kernels for rendering and output.
* Quarto supports better cross-referencing, LaTeX output, and rendering targets, but is slower in general to render large projects.
* JupyterBook is better integrated with `conda` making it easier to deploy on a custom server as a website, but Quarto offers more robust CLI options for non-interactive deployment.

For me, I think Quarto is better for my purposes because:

1)  Working with a large team with `.ipynb` format objects is not feasible on a Git system, due to the way the format stores metadata per cell.  JupyterNotebooks are an _output_ format, not an _input_ format for large projects.
2)  Flexibility and user interface is key, and Quarto's ability to adapt is much better; it's basically half-way between a standard Notebook and a full-blown static site generator, which is kind of perfect for me.
3)  LaTeX and BibTeX support with multiple render formats is essential for teaching and research; I can write nice-looking slides in RevealJS, and render handouts in LaTeX, along with textbook chapters or articles all from the same content sources without multiple programs or even significantly different code.

**Reply by jav-ed** [reply]
Thank you very much for your quick and detailed reply. I am also missing the feature of downloading .ipynb files directly from the webpage view and having access to the navigation bar.

**Reply by jlgraves-ubc** [reply]
Yeah - you can see the workaround we use to deploy on a JupyterHub on the website here:

https://comet.arts.ubc.ca/ 

It's not well-documented in the examples on the Quarto website, but you can also render to `.ipynb` if you have a `.qmd` format of a file, which can give you some options by using custom `_includes` in headers and stuff to pull in those files with a contextual link, but it will require some `lua` coding most likely or Javascript.

**Reply by jav-ed** [reply]
Thank you, Jonathan :)

**Reply by andrewpbray** [reply]
@jlgraves-ubc Thanks for the thorough breakdown of the distinctions as you see them - very helpful!
